### PR TITLE
MINIFICPP-2465 SplitContent

### PR DESCRIPTION
**Pull request by martinzink**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/standard-processors/processors/SplitContent.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  }
+}
+
+bool lastSplitIsEmpty(const std::vector<std::shared_ptr<core::FlowFile>>& splits) {
+  return splits.back()->getSize() != 0;
+}
+}  // namespace
+
+void SplitContent::onTrigger(core::ProcessContext& context, core::ProcessSession& session) {
+  gsl_Assert(!byte_sequence_.empty());
```
[the remark this answers is not in the document]
Comment (martinzink) [reply extensions/standard-processors/processors/SplitContent.cpp:117]: It was ensured implicitly because byte_sequence was a required property so it would have thrown if it was empty, but I've added an explicit check in onschedule aswell

File: extensions/standard-processors/tests/unit/SplitContentTests.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+  REQUIRE(original.size() == 1);
+  REQUIRE(splits.size() == 2);
+
+  auto expected_split_1 = createByteVector(5, 5, 5, 5);
+  auto expected_split_2 = createByteVector(1, 2, 3, 4);
+
+  CHECK(controller.plan->getContentAsBytes(*splits[0]) == expected_split_1);
+  CHECK(controller.plan->getContentAsBytes(*splits[1]) == expected_split_2);
+}
+
```
[the remark this answers is not in the document]
Comment (martinzink) [reply extensions/standard-processors/tests/unit/SplitContentTests.cpp]: In theory yes, I've tested it and it failed due to this edge case not being handled properly so awesome catch 👍 , fixed it in https://github.com/apache/nifi-minifi-cpp/pull/1872/commits/586e6f8079a6f8f0d4709b4cb4c3a31bef725fd1

**Review by szaszm — APPROVED** [review]
LGTM, with a caveat: I find the logic quite convoluted, but I didn't take the time to design something simpler that I could suggest.

If you have ideas to simplify the logic inside onTrigger and the related member functions, I think that would be a good target for refactoring.

File: libminifi/test/libtest/unit/TestBase.cpp [hunk 3]
```diff
@@ -672,11 +672,20 @@ void TestPlan::validateAnnotations() const {
   }
 }
 
+std::vector<std::byte> TestPlan::getContentAsBytes(const core::FlowFile& flow_file) const {
+  const auto content_claim = flow_file.getResourceClaim();
+  const auto content_stream = content_repo_->read(*content_claim);
+  const auto output_stream = std::make_shared<minifi::io::BufferStream>();
+  std::ignore = minifi::InputStreamPipe{*output_stream}(content_stream);
```
Comment (adamdebreceni) [inline libminifi/test/libtest/unit/TestBase.cpp:679]: despite its namespace `internal::pipe` is used in a number of places, so we could use that here as well (I think we should "graduate" it out of the internal namespace in the future)

File: extensions/standard-processors/processors/SplitContent.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      matching_bytes++;
+      if (matching_bytes == byte_sequence_.size()) {
+        // Found the Byte Sequence
+        finalizeLatestSplitContent(session, splits.back(), buffer);
+        ended_with_byte_sequence = true;
+        matching_bytes = 0;
+        buffer.clear();
+      }
+    } else {
+      matching_bytes = 0;
```
Comment (adamdebreceni) [inline extensions/standard-processors/processors/SplitContent.cpp:155]: this could fail for the following case: byte_sequence_ = "ab" and content = "aab", when the matching fails for the second 'a' it starts from the beginning for the next character

Comment (martinzink) [reply extensions/standard-processors/processors/SplitContent.cpp:183]: thanks, I've implemented your idea (talked about it in private), https://github.com/apache/nifi-minifi-cpp/pull/1872/commits/66435037a5a85eb381936ab3a0c708e44c13a73c

File: libminifi/test/libtest/unit/TestBase.cpp (hunk 3 above)
Comment (martinzink) [reply libminifi/test/libtest/unit/TestBase.cpp]: https://github.com/apache/nifi-minifi-cpp/pull/1872/commits/66435037a5a85eb381936ab3a0c708e44c13a73c

File: extensions/standard-processors/processors/SplitContent.h [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+   private:
+    struct node {
+      std::byte byte;
+      std::unordered_map<std::byte, size_type> cache;
+      std::optional<size_type> previous_max_match;
+    };
+    std::vector<node> byte_sequence_nodes_;
+    const std::vector<std::byte> byte_sequence_;
+  };
```
Comment (szaszm) [inline extensions/standard-processors/processors/SplitContent.h:127]: I think this class could be moved to the .cpp file

**Review by adamdebreceni — APPROVED** [review]

**Comment by martinzink** [comment]
I still found some discrepancies compared to NiFi, converting to draft while I iron out the kinks

File: extensions/standard-processors/processors/SplitContent.h (hunk 5 above)
Comment (martinzink) [reply extensions/standard-processors/processors/SplitContent.h]: We are using it as a member for SplitContent

**Review by szaszm — APPROVED** [review]

**Review by adamdebreceni — APPROVED** [review]
